### PR TITLE
Improvements to dataflow const-prop

### PR DESCRIPTION
**Pull request by cjgillot**
Partially cherry-picked from https://github.com/rust-lang/rust/pull/110719

r? @oli-obk 
cc @jachris

### CODE REVIEW

File: compiler/rustc_mir_transform/src/dataflow_const_prop.rs [hunk 1]
```diff
@@ -194,6 +211,19 @@ impl<'tcx> ValueAnalysis<'tcx> for ConstAnalysis<'_, 'tcx> {
...
+                if let ty::Array(_, len) = place_ty.ty.kind() {
+                    ConstantKind::Ty(*len)
+                        .eval(self.tcx, self.param_env)
+                        .try_to_scalar_int()
+                        .map_or(FlatSet::Top, FlatSet::Elem)
+                } else if let [ProjectionElem::Deref] = place.projection[..] {
+                    state.get_len(place.local.into(), self.map())
+                } else {
+                    FlatSet::Top
+                }
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/dataflow_const_prop.rs]: There's another option. If the slice is from a constant, we could even read the length from the constant, without it needing to be derived from an array.

File: compiler/rustc_mir_transform/src/dataflow_const_prop.rs [hunk 2]
```diff
@@ -193,47 +193,64 @@ impl<'tcx> ValueAnalysis<'tcx> for ConstAnalysis<'_, 'tcx> {
...
+                    FlatSet::Bottom => FlatSet::Bottom,
+                    FlatSet::Top => FlatSet::Top,
+                }
+            }
+            Rvalue::Cast(CastKind::Transmute, operand, _) => {
+                match self.eval_operand(operand, state) {
+                    FlatSet::Elem(op) => self.wrap_immediate(*op),
+                    FlatSet::Bottom => FlatSet::Bottom,
+                    FlatSet::Top => FlatSet::Top,
+                }
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/dataflow_const_prop.rs:252]: The use case is so rare, I'm not sure it's worth it.

File: compiler/rustc_mir_transform/src/dataflow_const_prop.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/dataflow_const_prop.rs]: In that case, we'd still have `Len(*_local)` where `_local` holds the slice. For that case, I'd rather directly implement general assignments `x = const` where the RHS is not a scalar.

File: compiler/rustc_mir_transform/src/dataflow_const_prop.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_mir_transform/src/dataflow_const_prop.rs:252]: not important, just saw 3 occurrences of this code snippet ^^

File: compiler/rustc_mir_dataflow/src/value_analysis.rs [hunk 3]
```diff
@@ -732,42 +731,52 @@ impl Map {
...
+            // Prepend new child to the linked list.
+            let len = self.places.push(PlaceInfo::new(Some(TrackElem::DerefLen)));
+            self.places[len].next_sibling = self.places[place].first_child;
+            self.places[place].first_child = Some(len);
+
+            let old = self.projections.insert((place, TrackElem::DerefLen), len);
+            assert!(old.is_none());
+
+            // Allocate a value slot since it doesn't have one.
+            assert!( self.places[len].value_index.is_none() );
```
Comment (klensy) [inline compiler/rustc_mir_dataflow/src/value_analysis.rs:773]: rustfmt skipped trailing spaces, but i can't repro this locally.

Comment (klensy) [reply compiler/rustc_mir_dataflow/src/value_analysis.rs:773]: Ah, let chains.

**Pull request merged**